### PR TITLE
sync: Remove readiness assertion in `watch::Receiver::changed()

### PR DESCRIPTION
**Pull request by zaharidichev**
## Motivation

After chatting with @hawkw I decided to try and tackle #2800. They pointed me to a[ bit of code](https://github.com/tokio-rs/tokio/pull/2814#discussion_r483794400) that might be useful when learning about how the `Notify` type works. I decided to play with loom and see how it works by writing a few tests. Then came around a problem with the `Watch` type. The current loom smoke test for `Watch` exercises only one receiver. Changing the test to have more than one will result in a panic: 

```
running 1 test
test sync::tests::loom_watch::smoke ... thread 'main' panicked at 'assertion failed: !res.is_ready()', tokio/src/sync/watch.rs:258:13
stack backtrace:
   0: std::panicking::begin_panic
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:505
   1: tokio::sync::watch::Receiver<T>::changed::{{closure}}::{{closure}}
             at ./src/sync/watch.rs:258
   2: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at ./src/future/poll_fn.rs:36
   3: tokio::sync::watch::Receiver<T>::changed::{{closure}}
             at ./src/sync/watch.rs:256
   4: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/mod.rs:79
   5: loom::future::block_on
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/loom-0.3.5/src/future/mod.rs:34
   6: tokio::sync::tests::loom_watch::smoke::{{closure}}
             at ./src/sync/tests/loom_watch.rs:18
   7: loom::model::Builder::check::{{closure}}
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/loom-0.3.5/src/model.rs:198
   8: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
   9: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1042
  10: loom::rt::scheduler::spawn_threads::{{closure}}::{{closure}}
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/loom-0.3.5/src/rt/scheduler.rs:140
  11: generator::gen_impl::GeneratorImpl<A,T>::init_code::{{closure}}
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/generator-0.6.21/src/gen_impl.rs:308
  12: generator::stack::StackBox<F>::call_once
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/generator-0.6.21/src/stack/mod.rs:135
  13: generator::stack::Func::call_once
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/generator-0.6.21/src/stack/mod.rs:117
  14: generator::gen_impl::gen_init::{{closure}}
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/generator-0.6.21/src/gen_impl.rs:513
  15: core::ops::function::FnOnce::call_once
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
  16: std::panicking::try::do_call
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:381
  17: ___rust_try
  18: std::panicking::try
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:345
  19: std::panic::catch_unwind
             at /Users/zaharidichev/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panic.rs:382
  20: generator::gen_impl::gen_init
             at /Users/zaharidichev/.cargo/registry/src/github.com-1ecc6299db9ec823/generator-0.6.21/src/gen_impl.rs:527
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
FAILED

failures:

failures:
    sync::tests::loom_watch::smoke

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 69 filtered out

error: test failed, to rerun pass '--lib'

```

This happens because in `watch::Receiver::changed` `Notified` was polled for the firt time to ensure that we register the waiter. The problem is that when there are more than one `Notified` instances tied to one `Notify`, it is possible for a `Notified` to be dropped without receiving its notification. If that happens this notification is left for another `Notified` instance to consume it. This means that it is not safe to assume that calling Notified::poll() for the first time shall always result in returning `Pending`, even if we are never calling `notify_one`.  

## Solution

We handle the case where polling the `Notified` future returns `Ready` right away.

Signed-off-by: Zahari Dichev <zaharidichev@gmail.com>

### CODE REVIEW

File: tokio/src/sync/watch.rs [hunk 1]
```diff
@@ -251,14 +251,22 @@ impl<T> Receiver<T> {
...
-            assert!(!res.is_ready());
-            Poll::Ready(())
-        })
-        .await;
+        // Polling the future here has dual purpose. The first one is to register
+        // the waiter so when `notify_waiters` is called it is notified. The second
+        // is to cover the case where another instance of `Notiified` has been dropped
+        // without receiving its notification. If that was the case polling the
+        // future for the first time will use this "lost" notification and return
+        // `Ready` immediatelly without registering any waiter
```
Comment (hawkw) [inline tokio/src/sync/watch.rs:259]: a couple typos and grammar nits:
```suggestion
        // Polling the future here has a dual purpose. The first one is to register
        // the waiter so that it is notified when `notify_waiters` is called. The second
        // is to cover the case where another instance of `Notified` has been dropped
        // without receiving its notification. If this has happened, polling the future
        // for the first time will use this "lost" notification and return  `Ready` 
        // immediately without registering any waiter.
```

File: tokio/src/sync/watch.rs [hunk 2]
```diff
@@ -251,14 +251,22 @@ impl<T> Receiver<T> {
...
+        let aquired_lost_notification =
+            crate::future::poll_fn(|cx| match Pin::new(&mut notified).poll(cx) {
+                Poll::Ready(()) => Poll::Ready(true),
+                Poll::Pending => Poll::Ready(false),
+            })
+            .await;
+
+        if aquired_lost_notification {
```
Comment (hawkw) [inline tokio/src/sync/watch.rs:267]: nit, take it or leave it: "acquired_lost_notification" is kind of a mouthful, what about something like
```suggestion
        let already_notified =
            crate::future::poll_fn(|cx| match Pin::new(&mut notified).poll(cx) {
                Poll::Ready(()) => Poll::Ready(true),
                Poll::Pending => Poll::Ready(false),
            })
            .await;

        if already_notified {
```
(also, "acquired" is spelled wrong in the code, so we should fix that even if we don't change the name :) )

**Review by hawkw — APPROVED** [review]
This looks right to me, thanks for the fix! @carllerche, anything I'm overlooking? 

I had a couple minor nits, but the change seems right.

**Comment by carllerche** [comment]
Ack, you are correct. There should be more tests :upside_down_face: thanks for catching this.

I don't think this will entirely solve the issue as we can concurrently send many updates and reproduce the problem.

As I see it, we have two options:

a) Update `Notify` to **not** resend the notification when `Notified` is dropped **if** the notification is from a `notify_waiters` call. In general, this would be a better behavior.

b) Implement a loop in `changed()` that awaits on `notified()` and checks if the version increased.

**Comment by zaharidichev** [comment]
@carllerche 

> a) Update Notify to not resend the notification when Notified is dropped if the notification is from a notify_waiters call. In general, this would be a better behavior.

That indeed makes the most sense to me as well. Just out of curiosity, how can I triger the problem after this change by sending concurrent updates. Can I have a loom test to exercise that part ?

**Comment by zaharidichev** [comment]
@hawkw I incorporated the suggestion to avoid resending the notifications if they have been trigerred by `notify_waiters`

**Review by carllerche — APPROVED** [review]
Looks good to me :+1: thanks. I'm not sure why CI is failing though. I will try restarting it.

File: tokio/src/sync/notify.rs [hunk 3]
```diff
@@ -105,6 +105,13 @@ pub struct Notify {
     state: AtomicU8,
     waiters: Mutex<WaitList>,
 }
+#[derive(Debug, Clone, Copy)]
+enum NotificationType {
+    // Notification trigerred by calling `notify_waiters`
+    AllWaiters,
+    // Notification trigerred by calling `notify_one`
+    OneWaiter,
```
Comment (hawkw) [inline tokio/src/sync/notify.rs:113]: typos:
```suggestion
    // Notification triggered by calling `notify_waiters`
    AllWaiters,
    // Notification triggered by calling `notify_one`
    OneWaiter,
```

File: tokio/src/sync/notify.rs [hunk 4]
```diff
@@ -583,13 +590,12 @@ impl Drop for Notified<'_> {
             }
 
             // See if the node was notified but not received. In this case, the
-            // notification must be sent to another waiter.
+            // notification must be sent to another waiter, only if it was
+            // triggered via `notify_one`
```
Comment (hawkw) [inline tokio/src/sync/notify.rs:594]: I might rephrase this to something like
```rust
            // See if the node was notified but not received. In this case, if
            // the notification was triggered via `notify_one`, it must be sent
            // to the next waiter.
```

File: tokio/src/sync/notify.rs [hunk 5]
```diff
@@ -105,6 +105,13 @@ pub struct Notify {
     state: AtomicU8,
     waiters: Mutex<WaitList>,
 }
+#[derive(Debug, Clone, Copy)]
+enum NotificationType {
```
Comment (hawkw) [inline tokio/src/sync/notify.rs:109]: nit: i'd add a newline here (surprised that rustfmt doesn't do that?)
```suggestion
}

#[derive(Debug, Clone, Copy)]
enum NotificationType {
```

**Review by hawkw — APPROVED** [review]
this looks right to me — i had a couple small style nits & typo corrections, but no issues with the implementation now that carl's comments have been addressed!
